### DISCUSSION TITLE
How to handle devices with additional information

### DISCUSSION THREAD

**Discussion by derjoerg** [kingsleyadam/local-abbfreeathome#34] [Q&A] [answered]
So I have a movement sensor, which also reports the current brightness level as one of the output datapoints. Here is the config
```
      "ABB700DA114D": {
        "deviceReboots": "49",
        "floor": "02",
        "room": "0B",
        "interface": "TP",
        "deviceId": "1008",
        "displayName": "Bewegungsmelder Flur zu WZ",
        "unresponsive": false,
        "unresponsiveCounter": 0,
        "defect": false,
        "channels": {
          "ch0000": {
            "floor": "02",
            "room": "0B",
            "displayName": "Bewegungsmelder Flur zu WZ",
            "selectedIcon": "52",
            "functionID": "11",   ### FID_MOVEMENT_DETECTOR
            "inputs": {
              "idp0000": {
                "pairingID": 256,   ### AL_INFO_ON_OFF
                "value": "0"
              }
            },
            "outputs": {
              "odp0000": {
                "pairingID": 6,   ### AL_TIMED_MOVEMENT
                "value": "0"
              },
              "odp0001": {
                "pairingID": 7,   ### AL_TIMED_PRESENCE
                "value": "1"
              },
              "odp0002": {
                "pairingID": 1027,   ### AL_BRIGHTNESS_LEVEL
                "value": "26.1"
              }
            },
            "parameters": {
              "par0034": "2",
              "par00d5": "20"
            }
          }
        },
        "parameters": {}
      },
```
What I found out so far:
- AL_INFO_ON_OFF: should be the state (but this is only set if the sensor is linked to a switch or a light in F@H)
- AL_BRIGHTNESS_LEVEL: can be later put in HA as additional attribute
- AL_TIMED_MOVEMENT: this is always 1, just in case of a movement it is additionally triggered, so perhaps in HA this should emit an Event??
- AL_TIMED_PRESENCE: I have no glue

How should a class look like to handle such a case? There are also other sensors/actors, which will have additional attribute (e.g. a wind sensor has the attribute wind-alarm, a RTC has things like valve position, current temperature and so on).

**Comment by derjoerg** [comment]
> So I have a movement sensor, which also reports the current brightness level as one of the output datapoints. Here is the config
> 
> ```
>       "ABB700DA114D": {
>         "deviceReboots": "49",
>         "floor": "02",
>         "room": "0B",
>         "interface": "TP",
>         "deviceId": "1008",
>         "displayName": "Bewegungsmelder Flur zu WZ",
>         "unresponsive": false,
>         "unresponsiveCounter": 0,
>         "defect": false,
>         "channels": {
>           "ch0000": {
>             "floor": "02",
>             "room": "0B",
>             "displayName": "Bewegungsmelder Flur zu WZ",
>             "selectedIcon": "52",
>             "functionID": "11",   ### FID_MOVEMENT_DETECTOR
>             "inputs": {
>               "idp0000": {
>                 "pairingID": 256,   ### AL_INFO_ON_OFF
>                 "value": "0"
>               }
>             },
>             "outputs": {
>               "odp0000": {
>                 "pairingID": 6,   ### AL_TIMED_MOVEMENT
>                 "value": "0"
>               },
>               "odp0001": {
>                 "pairingID": 7,   ### AL_TIMED_PRESENCE
>                 "value": "1"
>               },
>               "odp0002": {
>                 "pairingID": 1027,   ### AL_BRIGHTNESS_LEVEL
>                 "value": "26.1"
>               }
>             },
>             "parameters": {
>               "par0034": "2",
>               "par00d5": "20"
>             }
>           }
>         },
>         "parameters": {}
>       },
> ```
> 
> What I found out so far:
> 
> * AL_INFO_ON_OFF: should be the state (but this is only set if the sensor is linked to a switch or a light in F@H)
> * AL_BRIGHTNESS_LEVEL: can be later put in HA as additional attribute
> * AL_TIMED_MOVEMENT: this is always 1, just in case of a movement it is additionally triggered, so perhaps in HA this should emit an Event??
> * AL_TIMED_PRESENCE: I have no glue
> 
> How should a class look like to handle such a case? There are also other sensors/actors, which will have additional attribute (e.g. a wind sensor has the attribute wind-alarm, a RTC has things like valve position, current temperature and so on).

Also the parameters might come into play regarding additional attributes within HA. At the moment the parameters can only be modified in the F@H app. There is no (at least that I'm aware of) possibility to modify these through the REST-api. But even more worse, when you modify a parameter in F@H app it doesn't emits something in the websocket. This means e.g. for HA, that the values read out at the start of the integration are fixed. If you modify a parameter in the F@H app and you want to have this reflected in HA you need to reload everything. But perhaps this will change in the future.

**Comment by kingsleyadam** [comment]
To piggy back on the other discussions, this PR should include information about the devices and their purpose. https://github.com/kingsleyadam/local-abbfreeathome/pull/35

For this example I'd expect a new MovementDetector type of device that implements these specific datapoints as properties.

Within the home assistant integration we can tie it together, mapping this MovementDetector device to HomeAssistant entities.

**Comment by kingsleyadam** [comment]
From reading the [pairings](https://developer.eu.mybuildings.abb.com/fah_local/reference/pairingids) documentation I think `AL_TIMED_PRESENCE` would report movement regardless of the detectors configuration to report movement based on lux level.

The `AL_TIMED_MOVEMENT` pairing would only report movement if the configuration report movement (e.g. Movement when LUX Is higher than X).

I could be misunderstanding that though.

But I was also under the same impressions, `AL_TIMED_MOVEMENT` and probably `AL_TIMED_PRESENCE` should be created as [Event](https://developers.home-assistant.io/docs/core/entity/event) entities using the `EventDeviceClass.MOTION` class so it can be used in automations.

**Comment by kingsleyadam** [comment]
> Also the parameters might come into play regarding additional attributes within HA. At the moment the parameters can only be modified in the F@H app. There is no (at least that I'm aware of) possibility to modify these through the REST-api. But even more worse, when you modify a parameter in F@H app it doesn't emits something in the websocket. This means e.g. for HA, that the values read out at the start of the integration are fixed. If you modify a parameter in the F@H app and you want to have this reflected in HA you need to reload everything. But perhaps this will change in the future.

Yea, this is definitely an issue with the local api. I think for now we just have to ignore parameters in Home Assistant until it's feature complete in the local api. Hopefully it's just for configuration and not for device state.

**Comment by kingsleyadam** [comment, marked as answer]
@derjoerg, I've created a PR to add a movement detector device. I also have one of these so I was able to test it properly.

https://github.com/kingsleyadam/local-abbfreeathome/pull/40
